### ISSUE TITLE
Get global batch size for MirroredStrategy when using bucket_by_sequence_length

### ISSUE THREAD

**Issue by lioutasb** [tensorflow/tensorflow#29614]
**System information**
- TensorFlow version (you are using): 1.14-rc1
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**

I'm using [bucket_by_sequence_length](https://www.tensorflow.org/versions/r1.14/api_docs/python/tf/data/experimental/bucket_by_sequence_length) to batch sequences based on the maximum tokens allowed in a batch. This makes the batch size dynamic and not pre-determined. 

When I want to use distributed training using MirroredStrategy, I have to average the loss based on the global batch size or in other words the total examples I'm training in parallel in a single step across my GPUs. But since each GPU will have each own dynamic batch based on max tokens it's impossible for me to compute the total number of examples running in parallel since I only have access to a single GPU processing using the defined step_fn to run on [experimental_run](https://www.tensorflow.org/versions/r1.14/api_docs/python/tf/distribute/MirroredStrategy#experimental_run) function. 

It would be very helpful to be able to have access to global information as input to step_fn with details about all the batches running in parallel.

**Comment by guptapriya** [comment]
It might be possible to do this but I am not sure if the cost of synchronization to know the exact global batch size would be acceptable?
Instead of averaging over the exact global batch size, would it work to instead sum the loss, and then scale it by some "Appx global batch size" which is constant each step? Wouldn't that be a better way to weight your examples? Or do you necessarily want to have extra weight for examples when there are less examples per batch?

**Comment by byronyi** [comment]
That's an interesting question. Ping @lukaszkaiser if you have any comments or suggestions.

**Comment by lioutasb** [comment]
Okay, so the way that I'm solving this for now is as follows:

1) replace `experimental_run` with `experimental_run_v2` and  from TF1.14. This allows me to instead of giving the data iterator to the function and leave the `experimental_run` to fetch data, I will fetch them at each step and `experimental_run_v2` will take them as inputs to the `step_fn`.

2) By having the flexibility to give each distributed batch myself, I can now use `strategy.experimental_local_results` to the fetched batch, compute the total batch size across all GPUs and then give it as an extra parameters to the `step_fn`.

Of course, as @guptapriya said this has an extra cost that I have to pay due to synchronization of each GPU batch to be collected back to the master GPU in order to be computed. For now, it's not really a big problem for me as I don't see much of a latency (although I'm pretty sure it exists).

I think it would be easier to solve this problem if we had the option to choose instead of summing all the gradients from the GPUs, averaging them so that each can compute the the local gradients (based on the local batch). Of course, in the case of dynamic batch size it's still not exact but I think it's a better approximation that making a fixed approximation myself about the global batch size. I don't know though if this can be implemented with the Ring-AllReduce method that MirroredStrategy is using right now.

Disclaimer: I don't have much experience with distributed learning approaches so excuse me if some of my proposals don't make sense.

**Comment by guptapriya** [comment]
Could you share the skeleton code for your solution? I am not quite sure I completely followed it and want to make sure it would do what you expect.

Re: "option to choose instead of accumulating all the gradients from the GPUs, averaging them so that each can compute the the local gradients (based on the local batch)" -> I didn't understand this either. Currently : each replica computes it's gradients and then we add them. What are you proposing?

**Comment by lioutasb** [comment]
@guptapriya I can't share a full working example but what I did it is basically the following:

```
with strategy.scope():
    def train_step(inputs, total_batch_size):
        inp, tar = inputs
        with tf.GradientTape() as tape:
            predictions = model(inp, tar)
            cross_entropy = loss(predictions, tar)
            loss = tf.reduce_sum(cross_entropy) * (1.0 / tf.cast(total_batch_size, dtype=tf.float32))
            
        gradients = tape.gradient(loss, model.variables)
        update_vars = optimizer.apply_gradients(zip(gradients, model.variables))
        
        with tf.control_dependencies([update_vars]):
            return tf.identity(loss)
    
    def dist_train_step(dist_inputs):
        """
        dist_inputs will be a list of all the features (x, y) and each feature will be PerReplica object 
        """
        x_per_gpu_as_list = strategy.experimental_local_results(dist_inputs[0])
        batch_sizes = [tf.shape(x_gpu)[0] for x_gpu in x_per_gpu_as_list]
        total_batch_size = tf.reduce_sum(tf.stack(batch_sizes))
        
        per_replica_losses, per_replica_acc = strategy.experimental_run_v2(train_step, 
                                                                           args=(dist_inputs, 
                                                                                 total_batch_size))
        
        mean_loss = strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses, axis=None)
        
        return mean_loss
    
    train_iterator = train_dataset.make_initializable_iterator()
    train_iterator_init = train_iterator.initialize()
    
    dist_train = dist_train_step(train_iterator.get_next())
    
    with tf.Session(config=config) as sess:
        sess.run([v.initializer for v in all_variables])
        sess.run(train_iterator_init)
        
        step = 0
        loss_sum = 0
        while True:
            try:
                loss = sess.run([dist_train])
                loss_sum += loss
                acc_sum += acc
                
                if step % 100 == 0:
                    print(step, loss_sum/(step+1))
            
                step += 1
            except tf.errors.OutOfRangeError:
                print("Done!")
                break
```

**Comment by guptapriya** [comment]
Ah i see. Your code looks right to me, thanks for sharing! 

what would be a concrete API change to make this type of use case easy? It doesn't seem obvious given that we don't have a lot of insight into the user's inputs in experimental_run_v2. They may or may not have a batch dimension, for instance.
